filesystem/unix_permission.go

```diff
@@ -7,43 +7,99 @@ package filesystem
+	var stat unix.Stat_t
+	if err := unix.Stat(filePath, &stat); err != nil {
+		return 0, fmt.Errorf("cannot get file's stat %s: %v", filePath, err)
+	}
+	return uint32(stat.Mode), nil
+}
+
+func GetFileOwnerUserName(filePath string) (string, error) {
+	var stat unix.Stat_t
+	if err := unix.Stat(filePath, &stat); err != nil {
+		return "", fmt.Errorf("cannot get file's stat %s: %v", filePath, err)
+	}
+	if owner, err := user.LookupId(fmt.Sprintf("%d", stat.Uid)); err != nil {
+		return "", fmt.Errorf("cannot look up file owner's name %s: %v", filePath, err)
+	} else {
+		return owner.Username, nil
+	}
+}
+
+func GetFileGroupName(filePath string) (string, error) {
+	var stat unix.Stat_t
+	if err := unix.Stat(filePath, &stat); err != nil {
+		return "", fmt.Errorf("cannot get file's stat %s: %v", filePath, err)
+	}
+	if grp, err := user.LookupGroupId(fmt.Sprintf("%d", stat.Gid)); err != nil {
+		return "", fmt.Errorf("cannot look up file group name %s: %v", filePath, err)
+	} else {
+		return grp.Name, nil
+	}
+}
+
 // CheckFileRights check that the given file path has been protected by the owner.
 // If the owner is changed, they need at least the sudo permission to override the owner.
 func CheckFileRights(filePath string) error {
-	var stat syscall.Stat_t
-	if err := syscall.Stat(filePath, &stat); err != nil {
-		return fmt.Errorf("Cannot get file's stat %s: %v", filePath, err)
+	var stat unix.Stat_t
+	if err := unix.Stat(filePath, &stat); err != nil {
+		return fmt.Errorf("cannot get file's stat %s: %v", filePath, err)
 	}
 
 	// Check the owner of binary has read, write, exec.
-	if !(stat.Mode&(syscall.S_IXUSR) == 0 || stat.Mode&(syscall.S_IRUSR) == 0 || stat.Mode&(syscall.S_IWUSR) == 0) {
+	if !(stat.Mode&(unix.S_IXUSR) == 0 || stat.Mode&(unix.S_IRUSR) == 0 || stat.Mode&(unix.S_IWUSR) == 0) {
 		return nil
 	}
 
 	// Check the owner of file has read, write
-	if !(stat.Mode&(syscall.S_IRUSR) == 0 || stat.Mode&(syscall.S_IWUSR) == 0) {
+	if !(stat.Mode&(unix.S_IRUSR) == 0 || stat.Mode&(unix.S_IWUSR) == 0) {
 		return nil
 	}
 
-	return fmt.Errorf("File's owner does not have enough permission at path %s", filePath)
+	return fmt.Errorf("file's owner does not have enough permission at path %s", filePath)
 }
 
 // CheckFileOwnerRights check that the given owner is the same owner of the given filepath
 func CheckFileOwnerRights(filePath, requiredOwner string) error {
-	var stat syscall.Stat_t
-	if err := syscall.Stat(filePath, &stat); err != nil {
-		return fmt.Errorf("Cannot get file's stat %s: %v", filePath, err)
-	}
+	ownerUsername, err := GetFileOwnerUserName(filePath)
 
-	if owner, err := user.LookupId(fmt.Sprintf("%d", stat.Uid)); err != nil {
-		return fmt.Errorf("Cannot look up file owner's name %s: %v", filePath, err)
-	} else if owner.Name != requiredOwner {
-		return fmt.Errorf("Agent does not have permission to protect file %s", filePath)
+	if err != nil {
+		return fmt.Errorf("cannot look up file owner's name %s: %v", filePath, err)
```

Comment:
I have a general question about bubbling up errors in Golang.
If the called function returns an error, should the caller just return that error, or returned a modified/appended error message?

For example line 100 could have just been `return err`

Reply:
I had the same question actually. `Error handling and go` doc suggests to keep `error` separate from `message` and any other important information because `error` alone probably won't tell you enough. For example: 

```go
type appError struct {
    Error   error
    Message string
    Code    int
}
```

Reply:
I think that we should standardize how we format error messages in a separate PR. I agree with your callout